[MNPL - 16] Add joining and leaving existing lobbies

IdusMartiae opened the pull request.

IdusMartiae committed:
- Add joining lobbies
- Add leaving lobbies

Casper133 commented on src/main/controllers/LobbyController.ts:
```diff
@@ -32,7 +33,25 @@ export class LobbyController extends Controller {
+  @Delete('/lobby-participant')
+  @Security('JWT')
+  public async leaveLobby(@Request() request: RequestWithUser): Promise<void> {
+    logger.info(request.user.id);
```
зачем логгировать?

Casper133 commented on src/main/services/LobbyService.ts:
```diff
@@ -45,14 +62,6 @@ async function checkThatUserIsLobbyCreator(userId: number, lobbyId: number) {
   }
 }
 
-async function getLobbyByIdIfExists(lobbyId: number): Promise<Lobby> {
-  const lobby = await getRepository(Lobby).findOne(lobbyId);
-  if (lobby === undefined) {
-    throw new ClientError(LOBBY_DOES_NOT_EXIST);
-  }
-  return lobby;
-}
-
```
лучше оставить этот метод и использовать именно его, у него лучше название :)

Casper133 commented on src/main/services/LobbyService.ts:
```diff
@@ -80,10 +89,50 @@ export default {
+      throw new ClientError(LOBBY_REFUSED_PLAYER);
+    }
+
+    await getRepository(LobbyParticipant).save({
+      user: { id: userId },
+      lobby: { id: lobbyId },
+      isCreator: false,
+      isReady: false,
+    });
+
+    logger.info(
+      `User [USER_ID: ${userId}] joined lobby [LOBBY_ID: ${lobbyId}]`
+    );
+  },
+
+  async deleteLobbyParticipant(userId: number): Promise<void> {
+    const lobbyParticipant = await getRepository(LobbyParticipant).findOne({
+      user: { id: userId },
+    });
+
+    if (lobbyParticipant === undefined) {
+      throw new ClientError(USER_IS_NOT_A_LOBBY_MEMBER);
+    }
+
+    await getRepository(LobbyParticipant).delete(lobbyParticipant);
+    await logger.info(`User [USER_ID: ${userId}] left lobby`);
+  },
```
Ещё нужно добавить проверку на то, выходит ли из лобби его создатель. Если создатель, то также распускать лобби (об этом было в описании таски)

IdusMartiae replied:
Катя в своей таске сделала метод роспуска лобби создателем

Casper133 replied:
Да, но у неё реализация эндпоинта на конретно роспуск лобби. А здесь нужно сделать то же самое, но если выходит создатель из лобби. Это для того, чтобы фронт не задумывался, нужно ли делать дополнительно запрос на роспуск лобби или хватит только выйти из лобби, а бек-энд сам подчистит лобби

Casper133 commented on src/main/services/LobbyService.ts:
```diff
@@ -7,21 +7,37 @@ import LobbyCreationResponseDto from '../models/responses/LobbyCreationResponseD
-const NOT_AN_OWNER = 'Вы не являетесь создателем лобби';
+const MAX_PLAYERS_NUMBER = 6;
+
+const LOBBY_DOES_NOT_EXIST = 'Lobby does not exist';
+const LOBBY_REFUSED_PLAYER = 'Lobby refused connection';
 const LOBBY_CREATOR_DOES_NOT_EXIST = 'Создать лобби не определён';
 
+const USER_IS_A_LOBBY_MEMBER = 'User is a lobby member';
+const USER_IS_NOT_A_LOBBY_MEMBER = 'User is not a lobby member';
+const NOT_AN_OWNER = 'Вы не являетесь создателем лобби';
```
Какие-то сообщения на русском, какие-то на английском. Изначально у нас весь интерфейс на русском, поэтому сообщения также должны быть на русском

IdusMartiae committed:
Merge branch 'develop' into MNPL-16_Add_joining_and_leaving_lobbies

IdusMartiae committed:
fix

IdusMartiae committed:
Merge remote-tracking branch 'origin/MNPL-16_Add_joining_and_leaving_lobbies' into MNPL-16_Add_joining_and_leaving_lobbies

# Conflicts:
#	src/main/controllers/LobbyController.ts

IdusMartiae committed:
fix

IdusMartiae committed:
- Убрал лишний путь на Delete запрос: все пользователи выходят из лобби по Delete запросу к /lobbies/lobby-participant
- Изменил метод deleteLobby(): теперь метод удаляет лобби по указанному ID.
- Удалил лишнюю функцию проверки на создателя лобби
- Убрал одинаковые по смыслу константы с разными именами
- Теперь при Delete /lobby-participant удаляет лобби с юзерами, если создатель лобби вышел

IdusMartiae committed:
Убрал комментарий

Casper133 commented on src/main/services/LobbyService.ts:
```diff
@@ -91,15 +75,60 @@ export default {
+    }
+
+    if (lobbyParticipant.isCreator) {
+      await this.deleteLobby((await lobbyParticipant.lobby).id);
```
лишний await перед lobbyParticipant.lobby

IdusMartiae committed:
remove await
